mesonbuild/minstall.py

```diff
@@ -236,14 +239,31 @@ def restore_selinux_contexts() -> None:
         # If the list of files is empty, do not try to call restorecon.
         return
 
-    with subprocess.Popen(['restorecon', '-F', '-f-', '-0'],
-                          stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as proc:
-        out, err = proc.communicate(input=b'\0'.join(os.fsencode(f) for f in selinux_updates) + b'\0')
-        if proc.returncode != 0 and not os.environ.get('DESTDIR'):
-            print('Failed to restore SELinux context of installed files...',
-                  'Standard output:', out.decode(),
-                  'Standard error:', err.decode(), sep='\n')
+    proc, out, err = Popen_safe(['restorecon', '-F', '-f-', '-0'], (b'\0'.join(os.fsencode(f) for f in selinux_updates) + b'\0').decode())
+    if proc.returncode != 0 :
+        print('Failed to restore SELinux context of installed files...',
+              'Standard output:', out,
+              'Standard error:', err, sep='\n')
+
+def apply_ldconfig() -> None:
+    '''
+    Apply ldconfig to update the ld.so.cache.
+
+    If $DESTDIR is set, do not execute ldconfig.
+    '''
+    if not shutil.which('ldconfig'):
+        # If we don't have ldconfig, failure is ignored quietly.
+        return
+
+    if os.environ.get('DESTDIR'):
+        print('Do not execute ldconfig if DESTDIR is set')
```

Comment:
It seems odd to print a warning if we don't do anything due to DESTDIR, but not if we *can't* do anything because we cannot find ldconfig, and not a status message if we *do* do something.

Elsewhere, we do both, but we do them via `self.log()` which respects the "quiet" option.

Wording could be better, see e.g.:

```
self.log(f'Skipping custom install script because DESTDIR is set {name!r}')
self.log(f'Running custom install script {name!r}')
```

apply_ldconfig() should accept the destdir variable available in do_install, and, similar to self.run_install_script, check `if destdir:` to skip running ldconfig.

Reply:
> It seems odd to print a warning if we don't do anything due to DESTDIR, but not if we _can't_ do anything because we cannot find ldconfig, and not a status message if we _do_ do something.

We could be silent about DESTDIR too IMHO. If we don't have ldconfig it's probably a platform that does not use it, like Windows. I think it's fine to be silent about it.

> Elsewhere, we do both, but we do them via `self.log()` which respects the "quiet" option.

Would be nice to use `self.log()` but that function is outside the class, so would require a bit more of refactoring. It is already very unconsistent, there are tones of `print()`. IMHO this is not blocker for this PR, we are diverting already a lot from original goal.

> apply_ldconfig() should accept the destdir variable available in do_install, and, similar to self.run_install_script, check `if destdir:` to skip running ldconfig.

It's nicer to not depend on the env indeed. Note that it works fine because we have `os.environ['DESTDIR'] = self.options.destdir` but I agree better not count on it.

Reply:
A better check might be if islinux or isbsd, and then allow ldconfig to raise "command not found" errors.

The reason I suggested passing destdir around is indeed specifically because we read the env var once in a consistent manner and then pass that python variable around for reuse.

Reply:
On the self.log thing... I realized after your comment that this PR adds the ldconfig function to DirMaker which seems wrong. There's a reason restoring selinux is done in DirMaker because it's about low level file handling. But the ldconfig should he defined at the same level as running install scripts. So it should then have access to self.log()

Reply:
is it better this way ?

I dont think it can happen to have a linux/BSD without ldconfig installed